File: python/visio_schema/transport/link.py
"""fd byte I/O — the raw file-descriptor layer beneath endpoints.

There is no Link object: the fd IS the link. An endpoint owns one fd and does its
own non-blocking poll/read/write through these helpers; a reopenable endpoint gets
a fresh fd from an ``FdFactory`` (``Callable[[], int]``, -1 on failure) on each
reconnect. Mirrors the C++ ``link.hpp`` fd helpers.

Works for ``os.openpty()`` pairs (tests + the interop harness) and real serial
fds (``/dev/ttyUSB*`` / ``/dev/ttyGS0``) alike.
"""
from __future__ import annotations

import os
import termios
import tty
from collections.abc import Callable

# A source of fresh, connected fds for a reopenable endpoint (-1 on failure).
FdFactory = Callable[[], int]


def set_nonblocking(fd: int) -> None:
    """Set the fd non-blocking (so read_some/write_some never block)."""
    os.set_blocking(fd, False)


def set_raw_mode(fd: int) -> None:
    """Best-effort raw tty mode. No-op on non-ttys (termios refuses with ENOTTY)."""
    try:
        tty.setraw(fd, termios.TCSANOW)
    except termios.error:
        pass


def read_some(fd: int, max_bytes: int) -> bytes | None:
    """One non-blocking read. Returns the bytes read (len>0), ``b""`` on
    would-block (EAGAIN — no data yet, NOT EOF), or ``None`` on EOF / a dead fd."""
    try:
        chunk = os.read(fd, max_bytes)
    except BlockingIOError:
        return b""
    except OSError:
        return None
    if not chunk:
        return None  # EOF
    return chunk


def write_some(fd: int, data: bytes) -> int:
    """One non-blocking write. Returns bytes accepted (0..len), 0 on would-block
    (EAGAIN), or -1 on a broken fd. Mirrors ::write."""
    try:
        return os.write(fd, data)
    except BlockingIOError:
        return 0
    except (BrokenPipeError, OSError):
        return -1


def close_fd(fd: int) -> None:
    """Close ``fd`` (no-op on < 0). Flushes any queued TX first so a CDC-ACM gadget
    close doesn't block when no host is reading; no-op (ENOTTY) on non-ttys."""
    if fd < 0:
        return
    try:
        termios.tcflush(fd, termios.TCOFLUSH)
    except (termios.error, OSError):
        pass
    try:
        os.close(fd)
    except OSError:
        pass


def open_serial_fd(path: str) -> int:
    """Open a device path raw + non-blocking. Returns the fd or -1 on failure."""
    try:
        fd = os.open(path, os.O_RDWR | os.O_NOCTTY | os.O_CLOEXEC)
    except OSError:
        return -1
    set_raw_mode(fd)
    set_nonblocking(fd)
    return fd


def make_fd_pair() -> tuple[int, int]:
    """A connected pair of raw, non-blocking pty fds (master, slave). For tests
    and the cross-language interop harness: one end drives an endpoint, the other
    is read/written directly (or handed to the C++ peer by slave path)."""
    master, slave = os.openpty()
    for fd in (master, slave):
        set_raw_mode(fd)
        set_nonblocking(fd)
    return master, slave

File: cpp/src/transport/link.cc
#include "visio_schema/transport/link.hpp"

#include <arpa/inet.h>
#include <fcntl.h>
#if defined(__linux__)
#include <sys/prctl.h>
#endif
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <poll.h>
#include <sys/socket.h>
#include <termios.h>
#include <unistd.h>

#include <cerrno>

// MSG_NOSIGNAL (Linux) suppresses SIGPIPE per-send; macOS/BSD lack it and use the
// SO_NOSIGPIPE socket option instead (set by SetNoSigpipe at socket creation).
// Define it to 0 where absent so the send() call still compiles.
#ifndef MSG_NOSIGNAL
#define MSG_NOSIGNAL 0
#endif

namespace visio_schema::transport {

namespace {

// FD_CLOEXEC on `fd` — portable replacement for the SOCK_CLOEXEC socket() flag
// (Linux/BSD only). Best-effort; a missing CLOEXEC is not fatal.
void SetCloexec(int fd) {
  if (fd < 0) return;
  const int fl = ::fcntl(fd, F_GETFD, 0);
  if (fl >= 0) ::fcntl(fd, F_SETFD, fl | FD_CLOEXEC);
}

// Suppress SIGPIPE on writes to a socket whose peer hung up. Linux does this
// per-send via MSG_NOSIGNAL; macOS/BSD need the SO_NOSIGPIPE socket option. No-op
// where neither exists (the MSG_NOSIGNAL send path covers it).
void SetNoSigpipe(int fd) {
#ifdef SO_NOSIGPIPE
  int one = 1;
  ::setsockopt(fd, SOL_SOCKET, SO_NOSIGPIPE, &one, sizeof(one));
#else
  (void)fd;
#endif
}

}  // namespace

bool SetNonblocking(int fd) {
  const int fl = ::fcntl(fd, F_GETFL, 0);
  if (fl < 0) return false;
  return ::fcntl(fd, F_SETFL, fl | O_NONBLOCK) == 0;
}

long WriteSome(int fd, const std::uint8_t* data, std::size_t len) {
  for (;;) {
    // send(MSG_NOSIGNAL) so a write to a socket whose peer hung up returns EPIPE
    // instead of raising SIGPIPE (which would kill the process). Falls back to
    // write() for non-sockets (serial / pty / regular fd → ENOTSOCK).
    ssize_t n = ::send(fd, data, len, MSG_NOSIGNAL);
    if (n < 0 && errno == ENOTSOCK) n = ::write(fd, data, len);
    if (n >= 0) return static_cast<long>(n);
    if (errno == EINTR) continue;
    if (errno == EAGAIN || errno == EWOULDBLOCK) return 0;  // send buffer full
    return -1;  // EPIPE / EIO / ENODEV / ECONNRESET ...
  }
}

long ReadSome(int fd, std::uint8_t* buf, std::size_t max_bytes) {
  for (;;) {
    const ssize_t n = ::read(fd, buf, max_bytes);
    if (n > 0) return static_cast<long>(n);
    if (n == 0) return -1;  // EOF
    if (errno == EINTR) continue;
    if (errno == EAGAIN || errno == EWOULDBLOCK) return 0;  // no data yet
    return -1;  // dead fd
  }
}

void CloseFd(int fd) {
  if (fd < 0) return;
  // Drop any queued output before closing: a CDC-ACM gadget close (gs_close)
  // otherwise BLOCKS draining the TX buffer when no host is reading. No-op
  // (ENOTTY) on sockets / regular fds.
  ::tcflush(fd, TCOFLUSH);
  ::close(fd);
}

void SetRawMode(int fd) {
  termios tio{};
  if (::tcgetattr(fd, &tio) == 0) {  // ENOTTY on non-ttys → no-op
    ::cfmakeraw(&tio);
    ::tcsetattr(fd, TCSANOW, &tio);
  }
}

int OpenSerialFd(const char* path) {
  const int fd = ::open(path, O_RDWR | O_NOCTTY | O_CLOEXEC);
  if (fd < 0) return -1;
  SetRawMode(fd);
  if (!SetNonblocking(fd)) {
    ::close(fd);
    return -1;
  }
  return fd;
}

int DialTcpFd(const char* host, std::uint16_t port, int timeout_ms) {
  sockaddr_in addr{};
  addr.sin_family = AF_INET;
  addr.sin_port = htons(port);
  if (::inet_pton(AF_INET, host, &addr.sin_addr) != 1) return -1;

  const int fd = ::socket(AF_INET, SOCK_STREAM, 0);
  if (fd < 0) return -1;
  SetCloexec(fd);
  SetNoSigpipe(fd);

  if (timeout_ms > 0) {
    // Bounded connect: go non-blocking first, then poll(POLLOUT) up to timeout_ms —
    // a dropped SYN (peer not yet accepting) fails fast instead of blocking the
    // caller for the kernel SYN-retry window (~127 s).
    if (!SetNonblocking(fd)) { ::close(fd); return -1; }
    if (::connect(fd, reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) != 0) {
      if (errno != EINPROGRESS) { ::close(fd); return -1; }
      pollfd pfd{fd, POLLOUT, 0};
      if (::poll(&pfd, 1, timeout_ms) <= 0) { ::close(fd); return -1; }
      int err = 0;
      socklen_t len = sizeof(err);
      if (::getsockopt(fd, SOL_SOCKET, SO_ERROR, &err, &len) != 0 || err != 0) {
        ::close(fd);
        return -1;
      }
    }
  } else {
    if (::connect(fd, reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) != 0) {
      ::close(fd);
      return -1;
    }
    if (!SetNonblocking(fd)) { ::close(fd); return -1; }
  }
  int one = 1;
  ::setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &one, sizeof(one));
  ::setsockopt(fd, SOL_SOCKET, SO_KEEPALIVE, &one, sizeof(one));
  return fd;
}

std::pair<int, int> MakeFdPair() {
  int sv[2] = {-1, -1};
  if (::socketpair(AF_UNIX, SOCK_STREAM, 0, sv) != 0) return {-1, -1};
  for (int fd : sv) {
    SetNonblocking(fd);
    SetCloexec(fd);
    SetNoSigpipe(fd);
  }
  return {sv[0], sv[1]};
}

int OpenTcpListenSocket(std::uint16_t port) {
  const int fd = ::socket(AF_INET, SOCK_STREAM, 0);
  if (fd < 0) return -1;
  SetCloexec(fd);
  int one = 1;
  ::setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one));
  sockaddr_in addr{};
  addr.sin_family = AF_INET;
  addr.sin_addr.s_addr = htonl(INADDR_ANY);
  addr.sin_port = htons(port);
  if (::bind(fd, reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) != 0 ||
      ::listen(fd, /*backlog=*/4) != 0) {
    ::close(fd);
    return -1;
  }
  return fd;
}

int AcceptCloexec(int listen_fd) {
#if defined(__linux__)
  return ::accept4(listen_fd, nullptr, nullptr, SOCK_CLOEXEC);
#else
  const int fd = ::accept(listen_fd, nullptr, nullptr);
  if (fd >= 0) {
    SetCloexec(fd);
    SetNoSigpipe(fd);
  }
  return fd;
#endif
}

void SetCurrentThreadName(const char* name) {
#if defined(__linux__)
  ::prctl(PR_SET_NAME, name, 0, 0, 0);
#else
  (void)name;
#endif
}

}  // namespace visio_schema::transport
